frontend/include/chpl/resolution/scope-types.h

```diff
@@ -156,9 +140,102 @@ class BorrowedIdsWithName {
     // return the element just past the last element
     return last+1;
   }
+ public:
+  /**
+    Iterator that skips invisible entries from the list of borrowed IDs.
+   */
+  class BorrowedIdsWithNameIter : public std::iterator<ID, std::forward_iterator_tag> {
+    // To allow use of isIDVisible
+    friend class BorrowedIdsWithName;
+   private:
+    /** The borrowed IDs over which we are iterating. */
+    const BorrowedIdsWithName* ids;
+    /** The ID this iterator is pointing too. */
+    const OwnedID* currentId;
+
+    BorrowedIdsWithNameIter(const BorrowedIdsWithName* ids, const OwnedID* currentId)
+      : ids(ids), currentId(currentId) {
+      fastForward();
```

Comment:
I'd prefer something like `if (skipPrivateVisibilities) fastForward();` and then simplifying `fastForward()` appropriately.

Reply:
Doing this would mean exposing currently private state in BorrowedIdsWithName (specifically `skipPrivateVisibilities`) and spreading around logic for visibility checks (which is currently concentrated in `isIdVisible`). Unless we have a performance reason, I personally prefer the more encapsulated version.